### PR TITLE
media-video/mjpg-streamer:

### PR DESCRIPTION
**Pull request by ChaosEngine**
Closes: https://bugs.gentoo.org/724802
Closes: https://bugs.gentoo.org/665122
Package-Manager: Repoman 2.3.22 (portage-2.3.99)
Signed-off-by: Andrzej Pauli <andrzej.pauli@gmail.com>

### CODE REVIEW

File: media-video/mjpg-streamer/metadata.xml [hunk 1]
```diff
@@ -5,6 +5,10 @@
 		<email>andrzej.pauli@gmail.com</email>
 		<name>Andrzej Pauli</name>
 	</maintainer>
+	<maintainer type="person">
+		<email>aidecoe@gentoo.org</email>
+		<name>Amadeusz Żołnowski</name>
+	</maintainer>
```
Comment (juippis) [inline media-video/mjpg-streamer/metadata.xml:11]: Why are you adding him?

File: media-video/mjpg-streamer/mjpg-streamer-0_pre20200524-r1.ebuild [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+	output-file	output-udp +output-http output-autofocus output-rtsp output-viewer output-zmqserver
+	www http-management wxp-compat"
+REQUIRED_USE="|| ( ${INPUT_PLUGINS} )
+	|| ( ${OUTPUT_PLUGINS} )"
+
+RDEPEND="virtual/jpeg
+	input-uvc? ( media-libs/libv4l )
+	input-opencv? ( media-libs/opencv )
+	output-zmqserver? ( dev-libs/protobuf-c net-libs/zeromq )
+	input-ptp2? ( media-libs/libgphoto2 )"
```
Comment (juippis) [inline media-video/mjpg-streamer/mjpg-streamer-0_pre20200524-r1.ebuild:30]: Please sort in alphabetical order, input < output.

File: media-video/mjpg-streamer/mjpg-streamer-0_pre20200524-r1.ebuild [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+DEPEND="${RDEPEND}
+	input-testpicture? ( media-gfx/imagemagick )"
+
+S="${WORKDIR}/${PN}-${EGIT_COMMIT}/mjpg-streamer-experimental"
+
+src_prepare() {
+	sed -i -e "s|.*RPATH.*||g" CMakeLists.txt
```
Comment (juippis) [inline media-video/mjpg-streamer/mjpg-streamer-0_pre20200524-r1.ebuild:37]: Append with `|| die`.

File: media-video/mjpg-streamer/mjpg-streamer-0_pre20200524-r1.ebuild [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+	use wxp-compat && sed -i -e \
+		's|^add_feature_option(WXP_COMPAT "Enable compatibility with WebcamXP" OFF)|add_feature_option(WXP_COMPAT "Enable compatibility with WebcamXP" ON)|g' CMakeLists.txt
```
Comment (juippis) [inline media-video/mjpg-streamer/mjpg-streamer-0_pre20200524-r1.ebuild:39]: Open up properly with `if use; then`, append sed with `|| die`.

File: media-video/mjpg-streamer/mjpg-streamer-0_pre20200524-r1.ebuild [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+	local flag switch
+	for flag in ${IUSE_PLUGINS}; do
+		use ${flag} && switch='' || switch='#'
+		flag=${flag/input-/input_}
+		flag=${flag/output-/output_}
+		sed -i -e \
+			"s|^add_subdirectory(plugins\/${flag})|${switch}add_subdirectory(plugins/${flag})|" \
+			CMakeLists.txt
```
Comment (juippis) [inline media-video/mjpg-streamer/mjpg-streamer-0_pre20200524-r1.ebuild:48]: Needs `|| die`.

File: media-video/mjpg-streamer/mjpg-streamer-0_pre20200524-r1.ebuild [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+src_prepare() {
+	sed -i -e "s|.*RPATH.*||g" CMakeLists.txt
+	use wxp-compat && sed -i -e \
+		's|^add_feature_option(WXP_COMPAT "Enable compatibility with WebcamXP" OFF)|add_feature_option(WXP_COMPAT "Enable compatibility with WebcamXP" ON)|g' CMakeLists.txt
+
+	local flag switch
+	for flag in ${IUSE_PLUGINS}; do
+		use ${flag} && switch='' || switch='#'
+		flag=${flag/input-/input_}
+		flag=${flag/output-/output_}
```
Comment (juippis) [inline media-video/mjpg-streamer/mjpg-streamer-0_pre20200524-r1.ebuild:45]: Honestly I think this plugin management is just overcomplicating this ebuild. You can combine some of the flags, and enable everything that doesn't need external dependencies by default.

File: media-video/mjpg-streamer/mjpg-streamer-0_pre20200524-r1.ebuild [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+		sed -i -e \
+			"s|^add_subdirectory(plugins\/${flag})|${switch}add_subdirectory(plugins/${flag})|" \
+			CMakeLists.txt
+	done
+	use http-management && sed -i -e \
+		's|^add_feature_option(ENABLE_HTTP_MANAGEMENT "Enable experimental HTTP management option" OFF)|add_feature_option(ENABLE_HTTP_MANAGEMENT "Enable experimental HTTP management option" ON)|g' plugins/output_http/CMakeLists.txt
```
Comment (juippis) [inline media-video/mjpg-streamer/mjpg-streamer-0_pre20200524-r1.ebuild:51]: Open up properly, `if use; then`, append `sed` with `|| die`.

File: media-video/mjpg-streamer/mjpg-streamer-0_pre20200524-r1.ebuild [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+	dolib.so *.so
+
+	if use www ; then
+		insinto /usr/share/${PN}
+		doins -r www
+	fi
+
+	dodoc README.md TODO
+
+	sed -e "s|@LIBDIR@|$(get_libdir)/${PN}/$(get_libdir)|g" "${FILESDIR}/${PN}.initd" | newinitd - ${PN}
```
Comment (juippis) [inline media-video/mjpg-streamer/mjpg-streamer-0_pre20200524-r1.ebuild:69]: This `sed` call needs to happen in `src_prepare`.

File: media-video/mjpg-streamer/metadata.xml (hunk 1 above)
Comment (ChaosEngine) [reply media-video/mjpg-streamer/metadata.xml:11]: He was a supporter of this ebuild and helped me maintaining it.

File: media-video/mjpg-streamer/mjpg-streamer-0_pre20200524-r1.ebuild (hunk 6 above)
Comment (ChaosEngine) [reply media-video/mjpg-streamer/mjpg-streamer-0_pre20200524-r1.ebuild:45]: The thing is, that the outcome of build is main binary mjpg_streamer and plugin binaries. And not always all the plugin binaries are needed to be built/loaded. That's why I've given user ability to choose. Other flags are conditionals for compatibilities with weird devices or bug overcoming or simple optional features.

Comment (ChaosEngine) [reply media-video/mjpg-streamer/mjpg-streamer-0_pre20200524-r1.ebuild:48]: Now I can see that unfortunately this project is dropping support for media-libs/opencv-4.X.X and only supporting media-libs/opencv-3.x which is sadly not in portage anymore, so I will drop this use flag also. Others plugins and use flags to keep?

File: media-video/mjpg-streamer/metadata.xml (hunk 1 above)
Comment (thesamesam) [reply media-video/mjpg-streamer/metadata.xml:11]: yes

File: media-video/mjpg-streamer/mjpg-streamer-0_pre20200524-r1.ebuild (hunk 6 above)
Comment (thesamesam) [reply media-video/mjpg-streamer/mjpg-streamer-0_pre20200524-r1.ebuild:47]: Are there any plugins we can enable unconditionally? (Those not dragging in extra deps?)

And combine those which are logically similar.

The point is that exposing every single possible combination upstream isn't necessarily what we want to do in the ebuild -- it makes testing really hard.

File: media-video/mjpg-streamer/mjpg-streamer-0_pre20200524-r1.ebuild [hunk 9]
```diff
@@ -0,0 +1,88 @@
+# Copyright 1999-2020 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=7
+
+inherit eutils systemd
```
Comment (thesamesam) [inline media-video/mjpg-streamer/mjpg-streamer-0_pre20200524-r1.ebuild:6]: Where is eutils being used?

File: media-video/mjpg-streamer/mjpg-streamer-0_pre20200524-r1.ebuild [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+IUSE="input-testpicture input-control +input-file input-uvc input-http input-ptp2
+	output-file	output-udp +output-http output-autofocus output-rtsp output-viewer output-zmqserver
+	www http-management wxp-compat"
+REQUIRED_USE="|| ( ${INPUT_PLUGINS} )
+	|| ( ${OUTPUT_PLUGINS} )"
+
+RDEPEND="virtual/jpeg
+	input-uvc? ( media-libs/libv4l )
+	input-ptp2? ( media-libs/libgphoto2 )
+	output-zmqserver? ( dev-libs/protobuf-c net-libs/zeromq )"
```
Comment (thesamesam) [inline media-video/mjpg-streamer/mjpg-streamer-0_pre20200524-r1.ebuild:29]: Split onto new lines each:
```suggestion
	output-zmqserver? ( 
   	    dev-libs/protobuf-c 
	    net-libs/zeromq 
    )"
```

File: media-video/mjpg-streamer/mjpg-streamer-0_pre20200524-r1.ebuild [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+DEPEND="${RDEPEND}
+	input-testpicture? ( media-gfx/imagemagick )"
+
+S="${WORKDIR}/${PN}-${EGIT_COMMIT}/mjpg-streamer-experimental"
```
Comment (thesamesam) [inline media-video/mjpg-streamer/mjpg-streamer-0_pre20200524-r1.ebuild:33]:
```suggestion
S="${WORKDIR}/${PN}-${EGIT_COMMIT}/${PN}-experimental"
```

File: media-video/mjpg-streamer/mjpg-streamer-0_pre20200524-r1.ebuild [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+
+src_prepare() {
+	sed -i -e "s|.*RPATH.*||g" CMakeLists.txt || die
```
Comment (thesamesam) [inline media-video/mjpg-streamer/mjpg-streamer-0_pre20200524-r1.ebuild:36]: Normally we can use CMAKE_SKIP_RPATH in src_configure for this.

**Comment by ChaosEngine** [comment]
this also fixes https://bugs.gentoo.org/740514

File: media-video/mjpg-streamer/metadata.xml (hunk 1 above)
Comment (ChaosEngine) [reply media-video/mjpg-streamer/metadata.xml]: removed

File: media-video/mjpg-streamer/mjpg-streamer-0_pre20200524-r1.ebuild (hunk 12 above)
Comment (ChaosEngine) [reply media-video/mjpg-streamer/mjpg-streamer-0_pre20200524-r1.ebuild:39]: Well, I can't make it to work; looked at other ebuilds but failed.

File: media-video/mjpg-streamer/mjpg-streamer-0_pre20200524-r1.ebuild (hunk 6 above)
Comment (ChaosEngine) [reply media-video/mjpg-streamer/mjpg-streamer-0_pre20200524-r1.ebuild:50]: Yes, by default only USE="input-file* output-http" is enabled which is not pulling new deps. Only by enabling more/other plugins, it pulls more deps.

Comment (thesamesam) [reply media-video/mjpg-streamer/mjpg-streamer-0_pre20200524-r1.ebuild:50]: OK. But there are no others with no external deps which we could enable?

**Comment by thesamesam** [comment]
Please squash the commits and change to include your real name in the sign off.

File: media-video/mjpg-streamer/mjpg-streamer-0_pre20200524-r1.ebuild (hunk 6 above)
Comment (ChaosEngine) [reply media-video/mjpg-streamer/mjpg-streamer-0_pre20200524-r1.ebuild:50]: No, better set minimal requirements and let user's decide what to pull in. Currently I see#  qdepends mjpg-streamer
media-video/mjpg-streamer-0_pre20200524-r1: virtual/pkgconfig media-libs/libv4l virtual/jpeg

**Comment by expeditioneer** [comment]
Could you please add the ~arm64 keyword (i've tested this) to this ebuild? I think the raspberry pi users would also like to profit from this :+1: 

Could you also install the systemd service which is provided upstream?

Since you are refencing the group "video" should acct-group/video not be in RDEPEND?

**Comment by thesamesam** [comment]
re adding ~arm64, please file a bug for this or can @expeditioneer can verify that tests pass before you add it?

**Comment by expeditioneer** [comment]
this ebuild is working on my arm64 machine - sadly this project has no unit test, which could be executed

**Comment by ChaosEngine** [comment]
> Could you please add the ~arm64 keyword (i've tested this) to this ebuild? I think the raspberry pi users would also like to profit from this 👍

👍 
 
> Could you also install the systemd service which is provided upstream?

I am installing it: https://github.com/ChaosEngine/gentoo/blob/4faf3ab7c3b4df64ba5b1a8a7e02305a11ba274d/media-video/mjpg-streamer/mjpg-streamer-0_pre20200524-r1.ebuild#L80
Do you want me to change [it](https://github.com/jacksonliam/mjpg-streamer/blob/master/mjpg-streamer-experimental/mjpg_streamer%40.service) this [file](https://github.com/jacksonliam/mjpg-streamer/blob/master/mjpg-streamer-experimental/scripts/mjpg-streamer.service) ? 

> Since you are refencing the group "video" should acct-group/video not be in RDEPEND?

**Comment by expeditioneer** [comment]
@ChaosEngine I would prefer the[second](https://github.com/jacksonliam/mjpg-streamer/blob/master/mjpg-streamer-experimental/mjpg_streamer%40.service) service file linked.

**Comment by ChaosEngine** [comment]
> @ChaosEngine I would prefer the [second](https://github.com/jacksonliam/mjpg-streamer/blob/master/mjpg-streamer-experimental/mjpg_streamer%40.service) service file linked.

And it is exactly this which is included. Observe `emerge -B mjpg-streamer` and it's content:
http://sprunge.us/41GW9W

@expeditioneer can you try taking my branch and build the ebuild yourself to test this file is there?

**Comment by expeditioneer** [comment]
@ChaosEngine my bad, i have overlooked that. Sorry for the confusion.

**Comment by expeditioneer** [comment]
@ChaosEngine now i have a 'real' issue :-)
The shared object `input_raspicam.so` is not build because the [CMakeLists.txt](https://github.com/jacksonliam/mjpg-streamer/blob/85f89a8c321e799fabb1693c5d133f3fb48ee748/mjpg-streamer-experimental/plugins/input_raspicam/CMakeLists.txt) of this relies on the existence of the folder `/opt/vc/include` which does not necessarily have to be present on gentoo systems depending whether _media-libs/raspberrypi-userland_ or _media-libs/raspberrypi-userland-bin_ is installed.

IMHO the build of input_raspicam should also be controlled by a use flag and a conditional dependency on one of the two userland ebuilds.

**Comment by ChaosEngine** [comment]
> @ChaosEngine now i have a 'real' issue :-)
> The shared object `input_raspicam.so` is not build because the [CMakeLists.txt](https://github.com/jacksonliam/mjpg-streamer/blob/85f89a8c321e799fabb1693c5d133f3fb48ee748/mjpg-streamer-experimental/plugins/input_raspicam/CMakeLists.txt) of this relies on the existence of the folder `/opt/vc/include` which does not necessarily have to be present on gentoo systems depending whether _media-libs/raspberrypi-userland_ or _media-libs/raspberrypi-userland-bin_ is installed.
> 
> IMHO the build of input_raspicam should also be controlled by a use flag and a conditional dependency on one of the two userland ebuilds.

Ok, good one. Added  **input-raspicam** use and linked it with one of media-libs/raspberrypi-userland* packages.
Please try if this is working for you, thanks.

**Comment by ChaosEngine** [comment]
Thank You so much :-)

**Pull request closed without merging**